### ISSUE TITLE
Metric aucpr calculated wrongly with unbalanced data?

### ISSUE THREAD

**Issue by lorenzwalthert** [dmlc/xgboost#5681]
I use xgboost in a binary classification task with the metric `aucpr`. With balanced data, the xgboost metric is very similar to the one calculated by two other packages, but when the data is unbalanced, it's significantly different. Below, I wrote an R function that creates a test set and fits the models and estimates the area under the PR curve. Between the three runs demonstrated, all I varied was the distribution of the response variable (via sampling). I believe you probably assume that the classes are balanced when you compute the metric.

``` r
library(xgboost)

data(agaricus.train, package = "xgboost")
data(agaricus.test, package = "xgboost")

compare_aucpr <- function(prob) {
  dtrain <- xgb.DMatrix(agaricus.train$data, label = agaricus.train$label)
  dtest <- xgb.DMatrix(agaricus.test$data,
                       label = sample(
                         0:1, length(agaricus.test$label),
                         prob = prob, replace = TRUE
                       )
  )
  cat("this is the response distribution:", table(getinfo(dtest, "label")), "\n")
  
  watchlist <- list(train = dtrain, eval = dtest)
  
  ## A simple xgb.train example:
  param <- list(
    max_depth = 2, eta = 1, verbose = 0, nthread = 2,
    objective = "binary:logistic", eval_metric = "aucpr"
  )
  fit <- xgb.train(param, dtrain, nrounds = 2, watchlist)
  cat("\n\n")
  cat("xgboost gives: ", dplyr::last(fit$evaluation_log$eval_aucpr), "\n")
  composed <- tibble::tibble(pred = 1 - predict(fit, dtest), truth = factor(getinfo(dtest, "label")))
  
  cat("yardstick gives: ", yardstick::pr_auc(composed, truth, pred)$.estimate, "\n")
  pr <- PRROC::pr.curve(
    scores.class0 = composed$pred[composed$truth == "0"],
    scores.class1 = composed$pred[composed$truth == "1"],
    curve = T
  )
  cat("PRROC gives:", pr$auc.integral, "\n")
}

compare_aucpr(c(0.5, 0.5))
#> this is the response distribution: 798 813 
#> [1]  train-aucpr:0.930072    eval-aucpr:0.505099 
#> [2]  train-aucpr:0.971330    eval-aucpr:0.493634 
#> 
#> 
#> xgboost gives:  0.493634 
#> yardstick gives:  0.4941544 
#> PRROC gives: 0.486742
compare_aucpr(c(0.9, 0.1))
#> this is the response distribution: 1445 166 
#> [1]  train-aucpr:0.930072    eval-aucpr:0.112330 
#> [2]  train-aucpr:0.971330    eval-aucpr:0.107851 
#> 
#> 
#> xgboost gives:  0.107851 
#> yardstick gives:  0.9191467 
#> PRROC gives: 0.9081823
compare_aucpr(c(0.1, 0.9))
#> this is the response distribution: 174 1437 
#> [1]  train-aucpr:0.930072    eval-aucpr:0.897323 
#> [2]  train-aucpr:0.971330    eval-aucpr:0.892570 
#> 
#> 
#> xgboost gives:  0.89257 
#> yardstick gives:  0.1025903 
#> PRROC gives: 0.1061011
```

<sup>Created on 2020-05-20 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

The relevant section: xgboost gives 0.89 AUC, whereas the two other packages give ~0.10. This is completely different to me.

<sup>Created on 2020-05-19 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

I use the R package {xgboost} version 1.0.0.2.

**Comment by lorenzwalthert** [comment]
I think there are some problems in my example, will reopen if problem persists.
